### PR TITLE
Search

### PR DESCRIPTION
**Pull request by YuriHechter**
Main changes:
* Intents carousels based on grammar free text
* Typo suggest  based on grammar free text
* Refactoring, small fixes and some twicks with scores

https://docs.google.com/spreadsheets/d/1XTMp7wWflZCdOZNSAFeiI1p-V7Cyzg-rAdngS07OOGQ/edit#gid=81711526
checking eval is still in progress but looks good

### CODE REVIEW

File: search/engine.go [hunk 1]
```diff
@@ -607,6 +607,10 @@ func (e *ESEngine) DoSearch(ctx context.Context, query Query, sortBy string, fro
 	defer e.timeTrack(time.Now(), consts.LAT_DOSEARCH)
 
 	suggestChannel := make(chan null.String)
+	grammarsSingleHitIntentsChannel := make(chan []Intent, 1)
```
Comment (bbfsdev) [inline search/engine.go:610]: Add comment
// Initializing all channels.

File: search/engine.go [hunk 2]
```diff
@@ -800,7 +812,7 @@ func (e *ESEngine) DoSearch(ctx context.Context, query Query, sortBy string, fro
 				if hit.Score != nil {
 					if _, hasId := filtered.HitIdsMap[hit.Id]; hasId {
 						log.Infof("Same hit found for both regular and grammar filtered results: %v", hit.Id)
-						if hit.Score != nil {
+						if hit.Score != nil && *hit.Score > 5 {
```
Comment (bbfsdev) [inline search/engine.go:815]: Comment + Examples.

File: search/engine.go [hunk 3]
```diff
@@ -820,7 +832,7 @@ func (e *ESEngine) DoSearch(ctx context.Context, query Query, sortBy string, fro
 					}
 				}
 
-				boost := (*maxRegularScore * 0.9) / *filtered.MaxScore
+				boost := ((*maxRegularScore * 0.9) + 10) / *filtered.MaxScore
```
Comment (bbfsdev) [inline search/engine.go:835]: Comment + Examples

File: search/engine.go [hunk 4]
```diff
@@ -607,6 +607,10 @@ func (e *ESEngine) DoSearch(ctx context.Context, query Query, sortBy string, fro
 	defer e.timeTrack(time.Now(), consts.LAT_DOSEARCH)
 
 	suggestChannel := make(chan null.String)
+	grammarsSingleHitIntentsChannel := make(chan []Intent, 1)
+	grammarsFilterIntentsChannel := make(chan []Intent, 1)
+	grammarsFilteredResultsByLangChannel := make(chan map[string]FilteredSearchResult)
```
Comment (bbfsdev) [inline search/engine.go:612]: Missing channel size to be 1, othewise the "recover" phases will stack on this channel write as no one reads from this channel yet.

File: consts/consts.go [hunk 5]
```diff
@@ -463,6 +464,27 @@ var ES_INTENT_SUPPORTED_CONTENT_TYPES = map[string]bool{
 	CT_CLIP:                  true,
 }
 
+var INTENT_SUPPORTED_GRAMMAR_CT_VARIABLES = map[string]bool{
```
[the remark this answers is not in the document]
Comment (YuriHechter) [reply consts/consts.go]: done

File: consts/consts.go [hunk 6]
```diff
@@ -463,6 +464,27 @@ var ES_INTENT_SUPPORTED_CONTENT_TYPES = map[string]bool{
...
+	VAR_CT_PROGRAMS:    true,
+	VAR_CT_ARTICLES:    true,
+	VAR_CT_LESSONS:     true,
+	VAR_CT_BOOK_TITLES: true,
+}
+
+var INTENT_RT_BY_GRAMMAR_CT = map[string]map[string]bool{
```
[the remark this answers is not in the document]
Comment (YuriHechter) [reply consts/consts.go]: done

File: search/engine.go (hunk 1 above)
Comment (YuriHechter) [reply search/engine.go:611]: added

File: search/engine.go (hunk 2 above)
Comment (YuriHechter) [reply search/engine.go]: Comment added. Unfortunately cannot remember the example.

File: search/engine.go (hunk 3 above)
Comment (YuriHechter) [reply search/engine.go:836]: added

File: search/engine.go (hunk 4 above)
Comment (YuriHechter) [reply search/engine.go:613]: The recover is successful after panic even without setting channel size to  grammarsFilteredResultsByLangChannel

**Pull request merged**